src/plugins/apimon/apimon.cpp

```diff
@@ -0,0 +1,435 @@
+                    "\"ProcessName\": %s, "
+                    "\"UserName\": \"%s\", "
+                    "\"UserId\": %" PRIu64 ", "
+                    "\"PID\": %d, "
+                    "\"PPID\": %d, "
+                    "\"Method\": \"%s\", "
+                    "\"CalledFrom\": 0x%" PRIx64 ", "
+                    "\"ReturnValue\": 0x%" PRIx64 ", "
+                    "\"Arguments\": ",
+                    UNPACK_TIMEVAL(info->timestamp),
+                    escaped_pname,
+                    USERIDSTR(drakvuf), info->proc_data.userid,
+                    info->proc_data.pid, info->proc_data.ppid,
+                    info->trap->name,
+                    info->regs->rip,
+                    info->regs->rax);
+
+            print_arguments(drakvuf, info, ret_target->arguments, ret_target->argument_printers);
+            printf(", "
+                   "\"Extra\": {");
+            print_extra_data(extra_data);
+            printf("}}");
+            g_free(escaped_pname);
+            break;
+        default:
+        case OUTPUT_DEFAULT:
+            printf("[APIMON-USERHOOK] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 " ProcessName:\"%s\" UserID:%" PRIi64 " Method:\"%s\" CalledFrom:0x%" PRIx64 " ReturnValue:0x%" PRIx64 " Arguments:\"",
+            UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.name,
+            info->proc_data.userid, info->trap->name, info->regs->rax, info->regs->rip);
+            print_arguments(drakvuf, info, ret_target->arguments, ret_target->argument_printers);
+            printf("\"");
+            break;
+    }
+    printf("\n");
+
+    drakvuf_remove_trap(drakvuf, info->trap, nullptr);
+    delete ret_target;
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+static event_response_t usermode_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info* info)
+{
+    hook_target_entry_t* target = (hook_target_entry_t*)info->trap->data;
+
+    if (target->pid != info->proc_data.pid)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    vmi_v2pcache_flush(vmi, info->regs->cr3);
+
+    bool is_syswow = drakvuf_is_wow64(drakvuf, info);
+
+    access_context_t ctx =
+            {
+                    .translate_mechanism = VMI_TM_PROCESS_DTB,
+                    .dtb = info->regs->cr3,
+                    .addr = info->regs->rsp
+            };
+
+    bool success = false;
+    addr_t ret_addr = 0;
+
+    if (is_syswow)
+    {
+        uint32_t ret_addr_tmp;
+
+        if (vmi_read_32(vmi, &ctx, &ret_addr_tmp) == VMI_SUCCESS)
+        {
+            success = true;
+            ret_addr = ret_addr_tmp;
+        }
+    }
+    else
+    {
+        success = vmi_read_64(vmi, &ctx, &ret_addr) == VMI_SUCCESS;
+    }
+    drakvuf_release_vmi(drakvuf);
+
+    return_hook_target_entry_t* ret_target = new return_hook_target_entry_t();
+    drakvuf_trap_t* trap = new drakvuf_trap_t;
```

Comment:
I'm not really sure why you guys prefer using `new` instead of `g_try_malloc0`. What happens if you are low on memory and `new` fails? I have no idea where that exception would be caught in this case.

Reply:
Perhaps using `g_slice_new` would be a better route.

Reply:
fixed

Reply:
Right now there is a problem with not using `new` here, because constant references are involved in some structures. It is not possible to perform equivalent initialization using `g_slice_new` or `g_try_malloc0`:

```cpp
return_hook_target_entry_t* ret_target = new return_hook_target_entry_t(target->pid, target->plugin, target->argument_printers);
```

`argument_printers` is a `const&` to a vector and changing that would require some little refactor. I plan to do some refactor anyway, but not immediately now :P

Reply:
Alright, then perhaps do `std::nothrow` so you can check that the pointer is NULL or not.

Reply:
doen